### ISSUE TITLE
Allow subcycling for first solver when using method="first-participant"

### ISSUE THREAD

**Issue by BenjaminRodenberg** [precice/precice#1570]
**Please describe the problem you are trying to solve.**

Currently `advance(dt)` is used by the first participant to define the time window size, if `method="first-participant"` is used. Therefore, it is by the design of the API not possible to perform subcycling with the first participant at the same time.

**Describe the solution you propose.**

To be discussed. One possibility is to introduce a new API method for defining the time window size, if `method="first-participant"` is used. E.g. `precice.define_time_window_size(dt)`. However, we will have to think about some smart default values and appropriate error messages:

What if `method="first-participant"` is used and a user does not call `precice.define_time_window_size(dt)`? Use the `dt` of `precice.advance(dt)` or raise an error?

**Comment by uekerman** [comment]
I assume there is a motivating application example for this enhancement? Currently, I do not yet understand the user perspective.

How would the first participant compute the time window size? This sounds to me a lot like this participant would know that it is coupled (contradicting the black-box approach).

If we want this, a further complexity is the behavior of `define_time_window_size` (or `set_time_window_size`) in the different timesteps within one window.

**Comment by BenjaminRodenberg** [comment]
> I assume there is a motivating application example for this enhancement? Currently, I do not yet understand the user perspective.

The motivation for this is that we have to do subcycling, if we want to use higher order interpolation (#1523). This means: If the first participant wants to send a higher order interpolant to the second it cannot do this, if it is using `method="first-participant"`. I created the following figure for a case where `method="first-participant"` is used and the second participant uses subcycling:

![Screenshot from 2023-02-19 19-35-58](https://user-images.githubusercontent.com/5740604/219968090-d9870cb3-7195-4f9b-af09-7ed8ba128a34.png)

Imagine that we want to use $c_{\mathcal{F}i}, i \in {0,1,2,3}$ to create a cubic spline, because this has a positive effect on the time stepping of $\mathcal{S}$. Then we would need to use the time window size $\Delta T = \Delta t_\mathcal{F}^1 + \Delta t_\mathcal{F}^2 + \Delta t_\mathcal{F}^3$.

> How would the first participant compute the time window size? This sounds to me a lot like this participant would know that it is coupled (contradicting the black-box approach).

You are right. This part is tricky. I think from the perspective of $\mathcal{F}$ it makes not a lot of sense to define the size of a complete time window at any point in time, because this means that $\mathcal{F}$ either has to keep track of past time step sizes or has to know about future time step sizes.

> If we want this, a further complexity is the behavior of `define_time_window_size` (or `set_time_window_size`) in the different timesteps within one window.

Maybe a better idea, that also solves this problem: $\matchcal{F}$ calls a function `trigger_synchronization_now`? This function is only called once per time window (probably before the last `advance(dt)` of the substeps of $\mathcal{F}$). Then preCICE can take care of computing the time window size (adding up $\Delta t_\mathcal{F}^i$).

I can imagine different situations where $\mathcal{F}$ might want to call this function: 

1) after $n = 3$ timesteps, because it wants to be able to provide a cubic spline for $\mathcal{S}$, (again breaking the black-box to some degree)
2) if a certain event happens (crack, contact, mesh is recomputed...) or 
3) every $x$ seconds, but we want to give priority to the adaptive time stepping of $\mathcal{F}$ and not to the fixed `time-window-size` in the `precice-config.xml`. Meaning: We want to *roughly* couple every 1s, but it is also acceptable to do it after 1.1s, 10s or even later, if the adaptive time stepping of $\mathcal{F}$ decides that at 1s it is a bad idea to let a timestep end.

**Comment by fsimonis** [comment]
Suggestion: `define_next_time_window_size()`

Calling before `initialize()` sets the first timewindow size.
Calling before `advance()` sets the next timewindow size.

Communicate as with `method="first-participant"`.
Warn if the user overrides this in timesteps with different values:
```
WARNING: You have provided different values for the next timewindow size. preCICE will use the latest. Please use this only once per time window.
```

**Comment by uekerman** [comment]
> Calling before advance() sets the next timewindow size.

There would be multiple `advance` calls within one time window. That's the difficulty here.

> We can also leave this issue open until we see that this feature is really needed. At the moment it's a hypothetical and nice-to-have feature. Theoretically possible, but I don't know about a user that really needs it (yet).

Agree

**Comment by fsimonis** [comment]
> There would be multiple advance calls within one time window. That's the difficulty here.

If the API allows to specify the length of the **next** time window, then this information can be held onto by precice until the end of the current time window is reached.

The adapter knows when the time window is over via the result of initialize and advance. It thus know when it needs to provide the information at the latest. 

preCICE knows when the participant reaches the end of the time window and knows the requested next time window size. 

Once the end was reached, the participants exchange the next length of the time window, as we already do in serial schemes.

This should actually be easy to implement. (I know I know)

**Comment by BenjaminRodenberg** [comment]
## Some background

I realized yesterday that there is a problem when using waveform relaxation + `method="first-participant"` in certain situations:

Imagine you are performing several iterations (implicit coupling) on the same window. In each iteration the first participant $\mathcal{F}$ may define a different time window size for the current window. Let's say we have iteration $k$ and $k+1$, then $\Delta T^k$ and $\Delta T^{k+1}$ refer to the respective time window sizes. If $\Delta T^k < \Delta T^{k+1}$ this leads to the problem that $\mathcal{F}$ is allowed to sample from the waveform provided by the second participant $\mathcal{S}$for $t \in \left(\Delta T^k, \Delta T^{k+1}\right]$, but in the previous iteration $k$ the second participant $\mathcal{S}$ only computed a solution for $t \in \left[0, \Delta T^k\right]$.

One possible solution is to perform extrapolation, if the waveform is sampled for $t \in \left(\Delta T^k, \Delta T^{k+1}\right]$, but supporting extrapolation for waveforms is currently not planned. Another possibility is to only allow $\mathcal{F}$ to define the time window size $\Delta T$ in the first iteration $k=0$ of the window. In the following iterations $k>0$. Therefore, $\mathcal{F}$ will not be able to change this time window size anymore. (Limitation: What if $\mathcal{F}$ needs to reduce the time window size for stability reasons while iterating?)

## What does this have to do with this issue?

I think the following API might be a solution for both problems:

```python
precice.initialize()
precice.define_time_window_size(window_dt)  # sets time window size for first window
....
while precice.is_coupling_ongoing():
    ... # usual stuff
    precice_dt = precice.advance(dt)  # checks whether precice.define_time_window_size was called in this window, similar to precice.requires_writing_checkpoint etc.
    dt = min(precice_dt, solver_dt)

    if precice.is_time_window_complete():
        precice.define_time_window_size(window_dt)  # sets time window size for the following windows
```

Alternative:

```python
precice.initialize()
....
while precice.is_coupling_ongoing():
    if requires_writing_checkpoint():
        # checkpointing AND
        precice.define_time_window_size(window_dt)  # sets time window size for the following windows

    ... # usual stuff
    precice_dt = precice.advance(dt)  # 
    dt = min(precice_dt, solver_dt)
    ...
```

Or my favorite solution:

```python
precice.initialize()
....
while precice.is_coupling_ongoing():
    if requires_time_window_size():  # again, very similar to precice.requires_writing_checkpoint etc.
        precice.define_time_window_size(window_dt)  # sets time window size for the following windows

    ... # usual stuff
    precice_dt = precice.advance(dt)  # 
    dt = min(precice_dt, solver_dt)
    ...
```

**Comment by fsimonis** [comment]
At the risk of repeating myself: my suggestion it to prescribe the **next** time window size.

## Example

```py
### Initial time window
# request a size of the initial time-window 
precice.prescribe_next_time_window_size(initial_Dt)
precice_dt = precice.initialize() # preCICE now knows the initial Dt
# precice_dt is initial_Dt if method=first-participant and this participant is actually the first one

while precice.is_coupling_ongoing():
    ### Time windows during the coupling
    # ...
    preferred_dt = ...
    if (preferred_dt < precice_dt):
        next_preferred_DT = ...
        precice.prescribe_next_time_window_size(next_preferred_DT)
    dt = min(precice_dt, solver_dt)
    precice_dt = precice.advance(dt) # At the end, preCICE nows the next Dt 
    # ...
```


So calling `prescribe_next_time_window_size(preferred_Dt)`:
* merely requests a size for the next time window
* saves the Dt in the SolverInterface until `initialize` is called. It then uses it as the initial time window size.
* saves the Dt in the SolverInterface until `advance` reaches the end of the time window. It then uses it as the next time window size.

Instead of a requirement, one could also return weather preCICE will respect the request:

```py
accepted = prescribe_next_time_window_size(preferred_DT)
if (accepted):
  print("preCICE ignored me")
assert accepted, "I need this time window size"
```

**Comment by BenjaminRodenberg** [comment]
Fyi: I was close to creating a duplicate issue of this one since this feature was again requested at the preCICE workshop 2024.

**Comment by BenjaminRodenberg** [comment]
Discussed a possible implementation with @fsimonis.

From the perspective of first participant:

```python
while participant.is_coupling_ongoing():
    ...  # checkpointing, read, solve, write
    if participant.defines_window_size()  # safeguard to ensure participant is allowed to set time window size (condition: serial coupling, first participant, method="first-participant")
        and needs_to_perform_more_steps_in_this_window():  # decided by solver
        participant.one_more_step()  # will turn the next advance into a subcycling advance; throws exception if !participant.defines_window_size()
    else:
        pass  # next advance is the final advance in this window. Window size is the sum of all dt's collected in this window
    
    participant.advance(dt)  # concludes step or window (depending on whether precice.one_more_step() was called)
```

* :+1: This API should lead to a backwards compatible API. 
* :+1: We also do not need a lot of additional logic in advance.
* :-1: We have to check if the method `one_more_step()` may be called or not. For this purpose we need to offer the safeguard `defines_window_size()`. This reminds me a bit of the `is_required`/`perform_action` API that we just removed in v3. But @fsimonis and I also could not come up with a better solution here that is still descriptive.

**Comment by uekerman** [comment]
> I was close to creating a duplicate issue of this one since this feature was again requested at the preCICE workshop 2024.

For what use case exactly? If it is "only" higher order, wouldn't

```xml
method="first-participant" synchronize-every-n-time-windows="3"
```

still be the easier solution?

**Comment by fsimonis** [comment]
Different use-case but same issue is the use-case from the Michael Breuer's group at HSU during the preCICE workshop 2025.

They have a solver with a tuned adaptive time-step length that cannot be altered, so they need to use a serial scheme with first-participant.
It is then coupled to a code that can do significantly bigger time-steps.
The second participant is now limited by the tiny time-steps prescribed by the first participant.

Hence, they need some functionality to merge the small time-steps into a larger time-window.

This could be by amount like https://github.com/precice/precice/issues/1570#issuecomment-2397576748 
```xml
<time-window-size method="first-participant" synchronize-every-n-time-windows="3" />
```
or by size like
```xml
<time-window-size method="first-participant" min-time-window-size="1.0" />
```
.

Both seem very easy to misuse in the configuration though. Good error messages are important as usual.
